packages/ggit/src/fetch.ts

```diff
@@ -0,0 +1,81 @@
+import FormData = require('form-data');
+import fetch from 'node-fetch';
+import type { HeadersInit, RequestInit } from 'node-fetch';
+
+import type {
+  ApiCall,
+  Endpoint,
+  HttpMethod,
+  Method,
+  Uri,
+  Url,
+  TokenApiResponse,
+} from './index';
+
+export interface Options extends RequestInit {
+  method: HttpMethod;
+  redirect: 'follow';
+  headers: HeadersInit & {
+    authorization?: `Bearer ${TokenApiResponse['jwtToken']}`;
+    'Content-Type'?: 'application/json';
+  };
+  body: string | FormData;
+}
+
+export class Fetch<ApiType extends ApiCall> {
+  #url: Url = 'https://api.soilmetrics.eco/api';
+
+  #method: Method;
+
+  #endpoint: Endpoint;
+
+  #httpMethod: Options['method'] = 'POST';
+
+  constructor({ method, endpoint }: { method: Method; endpoint: Endpoint }) {
+    this.#method = method;
+    this.#endpoint = endpoint;
+  }
+
+  get uri(): Uri {
+    return `${this.#url}/${this.#method}/${this.#endpoint}` as const;
+  }
+
+  async fetch({
+    body,
+    headers,
+  }: {
+    body: ApiType['request']['body'];
+    headers: ApiType['request']['headers'];
+  }): Promise<ApiType['response']> {
+    try {
+      const options = this.buildOptions({
+        body,
+        headers,
+      });
+      const response = await fetch(this.uri, options);
+      const data = await response.json();
```

Comment:
This should probably go after the `response.ok` check unless you're passing `Accept: application/json` and their API is better behaved that most and actually returns JSON formatted bodies for all error codes.

Reply:
good call